iree/turbine/kernel/wave/wave.py

```diff
@@ -648,11 +649,13 @@ def _trace_and_get_kernel_signature(
 
         # Add grid and block dims to kernel launch info.
         dynamic_symbols_map = kwargs.get("dynamic_symbols_map", {})
-        try_convert_to_int = lambda x: int(x) if isinstance(x, sympy.Integer) else x
-        kernel_launch_info.grid = [
-            try_convert_to_int(safe_subs(x, dynamic_symbols_map))
-            for x in self.grid_type.dims
-        ]
+        # Convert the grid into a lambda that we can use to compute the grid dimension.
+        kernel_launch_info.grid = sympy.lambdify(
+            [list(dynamic_symbols_map.keys())], self.grid_type.dims
+        )
+        kernel_launch_info.grid_str = lambdastr(
```

Comment:
is grid_str mostly for caching purposes?

Reply:
Yes and also helps from a debuggability perspective because you can see the lambda function.

Reply:
makes sense! thanks :)

Reply:
Seems like most of the time we are rematerializing it with `kernel_info_str["grid"] = eval(kernel_info_str["grid_str"])`. Is it possible to refactor to get rid of `kernel_info_str["grid"] ` and `kernel_launch_info.grid` by opting to do a local `grid = eval(kernel_info_str["grid"])` inside `invoke_with_wave_runtime`

Reply:
There are two separate use cases. In the case where there's no caching, it does not make sense to construct a string and then convert the string to a lambda when we are about to launch the kernel, because we can directly convert the sympy expression into a lambda function. When we have to cache, then it makes sense to use the string representation since we can use json to store it and inspect it. In that situation, we have to use eval to convert a string into a lambda but luckily it's only a one time cost. In general I don't think it makes sense to eval every single time we want to run the kernel as that is an expensive operation.